pgjdbc/src/main/java/org/postgresql/jdbc/PgArray.java

```diff
@@ -310,14 +316,43 @@ private PgArrayList buildArrayList(String fieldString) throws SQLException {
     }
   }
 
+  private TimestampUtils getTimestampUtils() {
+    if (timestampUtils == null) {
+      final BaseConnection connection = getConnection();
+      timestampUtils = new TimestampUtils(!connection.getQueryExecutor().getIntegerDateTimes(), (Provider<TimeZone>) new QueryExecutorTimeZoneProvider(connection.getQueryExecutor()));
+    }
+    return timestampUtils;
+  }
+
   /**
    * Convert {@link ArrayList} to array.
    *
    * @param input list to be converted into array
    */
-  private Object buildArray(ArrayDecoding.PgArrayList input, int index, int count) throws SQLException {
+  private Object buildArray(ArrayDecoding.PgArrayList input, int index, int count, @Nullable Map<String, Class<?>> map) throws SQLException {
     final BaseConnection connection = getConnection();
-    return ArrayDecoding.readStringArray(index, count, connection.getTypeInfo().getPGArrayElement(oid), input, connection);
+    Object @Nullable [] array = (Object @Nullable []) ArrayDecoding.readStringArray(index, count, connection.getTypeInfo().getPGArrayElement(oid), input, connection);
+
+    if (map != null && array != null) {
+      for (int ii = 0; ii < array.length; ii++) {
+        @Nullable Object obj = array[ii];
+        if (obj instanceof PGobject) {
+          PGobject pgobj = (PGobject) obj;
+          Class<?> type = map.get(pgobj.getType());
+          if (type != null && SQLData.class.isAssignableFrom(type)) {
+            String value = pgobj.getValue();
+            Object result = new SQLDataReader().read(value, type, connection, getTimestampUtils());
+            if (result == null) {
+              array[ii] = null;
```

Comment:
This fails because `array` is declared as an array holding non-nullable, so the checker forbids assigning `null` there.

If you need nullable array of nullable objects, then you need to declare it as `@Nullable Object @Nullable [] array`

---

Here's a quick tip: https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md#null-safety

```java
  // array and elements: non-nullable
  String[] x;

  // array: nullable, elements: non-nullable
  String @Nullable [] x;

  // array: non-nullable, elements: nullable
  @Nullable String[] x;

  // array: nullable, elements: nullable
  @Nullable String @Nullable [] x;
```

Reply:
Ah thanks, I had that backward. I tried it the other way, but it was late, and I think I confused the error I got after that with a problem with this line. Anyway, thanks, resolved now.